### ISSUE TITLE
Restore fails if dependency isn't found on first source.

### ISSUE THREAD

**Issue by phillip-haydon** [NuGet/Home#7035]
If a project has more than 1 feed, restore fails on the first failure of a feed, it doesn't check all feeds before failing.

Right now I have to use `--ignore-failed-sources` which seems wrong, I want it to fail if the dependency isn't found in any source. 

If source 1 fails, it should attempt source 2. But it doesnt't... 

## Details about Problem

NuGet product used (NuGet.exe | VS UI | Package Manager Console | dotnet.exe): dotnet.exe

NuGet version (x.x.x.xxx): ??

dotnet.exe --version (if appropriate): 2.1.300

VS version (if appropriate): b/a

OS version (i.e. win10 v1607 (14393.321)): Windows Server 2016

Worked before? If so, with which NuGet version: Only happens with .NET Core related projects built from command line...

## Detailed repro steps so we can see the same problem

1. Have a project with dependency from NuGet, and a dependency from somewhere else that is not on NuGet (private feed) 

2. Try to restore

...

<details>
  <summary>Logs Snippet</summary>
<pre>
[04:18:37]Step 1/1: Build with .NET Core (.NET CLI (dotnet)) (4s)
[04:18:38][Step 1/1] dotnet.exe build My.Project.sln @D:\BuildAgent\temp\agentTmp\df1fc56684484103866a221f8ebd3c2f.rsp
[04:18:38][Step 1/1] build (4s)
[04:18:38][build] Starting: "C:\Program Files\dotnet\dotnet.exe" build My.Project.sln @D:\BuildAgent\temp\agentTmp\df1fc56684484103866a221f8ebd3c2f.rsp
[04:18:38][build] in directory: D:\BuildAgent\work\c9fff8a982ce155e
[04:18:38][build] Microsoft (R) Build Engine version 15.7.179.6572 for .NET Core
[04:18:38][build] Copyright (C) Microsoft Corporation. All rights reserved.
[04:18:38][build] 
[04:18:41][build]   Restoring packages for D:\BuildAgent\work\c9fff8a982ce155e\My.Project.Common.Tests\My.Project.Common.Tests.csproj...
[04:18:41][build]   Retrying 'FindPackagesByIdAsyncCore' for source 'https://teamcity.my-internal-site.net/httpAuth/app/nuget/v1/FeedService.svc/FindPackagesById()?id='NUnit'&semVerLevel=2.0.0'.
[04:18:41][build]   Response status code does not indicate success: 401 ( ).
[04:18:42][build]   Retrying 'FindPackagesByIdAsyncCore' for source 'https://teamcity.my-internal-site.net/httpAuth/app/nuget/v1/FeedService.svc/FindPackagesById()?id='NUnit'&semVerLevel=2.0.0'.
[04:18:42][build]   Response status code does not indicate success: 401 ( ).
[04:18:42][build] C:\Program Files\dotnet\sdk\2.1.300\NuGet.targets(114,5): error : Failed to retrieve information about 'NUnit' from remote source 'https://teamcity.my-internal-site.net/httpAuth/app/nuget/v1/FeedService.svc/FindPackagesById()?id='NUnit'&semVerLevel=2.0.0'.
[04:18:42][build] C:\Program Files\dotnet\sdk\2.1.300\NuGet.targets(114,5): error :   Response status code does not indicate success: 401 ( ).
[04:18:42][build] 
[04:18:42][build] Build FAILED.
[04:18:42][build] 
[04:18:42][build]     0 Warning(s)
[04:18:42][build]     1 Error(s)
[04:18:42][build] 
[04:18:42][build] Time Elapsed 00:00:02.83
[04:18:42][build] 
[04:18:42][build] 
</pre>
</details>

**Comment by nkolev92** [comment]
I assuming you are using Package Reference?

This is by design, the option to ignore failing sources on restore exists to allow using only the available sources. Ignoring failing sources by default is dangerous since a source could go down and a restore could pass and end up with different packages.


Please refer to https://github.com/NuGet/Home/issues/5607

**Comment by phillip-haydon** [comment]
@nkolev92 you’re missing the point. Package restore is broken by design.

It is impossible to restore if you use multiple sources and the package does not exist in 1 source.

Obviously ignoring is dangerous. But there’s no other work around because the current design is broken.

You should fail only if the package is found in no source, not the first source.

Please reopen my ticket as it’s still valid.

**Comment by nkolev92** [comment]
`--ignore-failed-sources` makes no impact on whether a package is found in a feed or not. 
NuGet does **not** fail if a package is not found in one feed but is in another. 
That's a very common scenario. 

That should **only** be used if a feed itself is inaccessible. 

Can you post the logs from your restore?

reopening until we get clarity here.

**Comment by phillip-haydon** [comment]
I posted the logs in the original issue. (Logs snippet at the bottom)

So the setup is basically:

We have Team City build our projects, TC has an internal NuGet feed which we can publish packages to. .NET Core seems to keep track of all sources (I'm not sure how it does this, but it appears to apply all sources to all builds once a source is used once, by this I mean when I created the project for .net core, it installed NUnit fine, the 2nd time I ran the build it failed as it now knew about the team city source)

What that means is, if I create a new project, and I reference something from nuget.org, then I push the changes to be built, .NET Core looks at all package sources. 

In the logs above the first source is:

https://teamcity.my-internal-site.net/httpAuth/app/nuget/v1/FeedService.svc/FindPackagesById()?id='NUnit'&semVerLevel=2.0.0

Because it does not find NUnit here, it fails, since this source only contains our internal dependencies.

If I set the flag `--ignore-failed-sources`

It prevents the hard fail, then looks at nuget.org and the local cache, and installs the dependency.

(If you can tell me how to clear the cache then I can run the build with and without to give you full logs)

I'm hoping I make sense.

**Comment by nkolev92** [comment]
Oh, I missed the log snippets, sorry about that. 

As I mentioned above. All feeds have to be accessible in PackageReference for a restore to happen. This is by design as having different sources available can lead to drastically different resolution graphs. 

The logs seem to indicate that you got a 401 from the private feed. 

The problem is not that the package is not available on the feed. Rather it's, I can't connect to the feed. 
A feed that a user can't connect to is considered unreachable, hence fails the restore. 

Next, NuGet firstly looks at the global packages folder (it's really an install location, I wouldn't call it a cache), before it goes to the online sources. 
If those packages are available in the global packages folder, it will use them. 
This is not as straightforward because NuGet will go online if there are imperfect dependency graphs or inexact versions. 
Namely "*" will cause NuGet to go online on the first restore. 
"2.0.0" while only having "2.5.0" in the version will do the same.  

Please refer to the below doc for information on what happens when NuGet installs a package. 
https://docs.microsoft.com/en-us/nuget/consume-packages/ways-to-install-a-package#what-happens-when-a-package-is-installed

tl;dr;

I think the original assessment still stands. 

All your sources need to available and this is by design. 

Does this make sense?

**Comment by phillip-haydon** [comment]
Hah. You missed that I attached logs. I missed the fact it says 401.

There’s a few things I’m confused about. (We are moving to .net core so still bits to figure out)

1) the first restore of the project was fine as it found nunit in nuget.org, built a nuget and pushed it to the internal feed.

When we built the project the second time it failed because it looked at our internal feed. 

How can I specify the feeds for a single project so I don’t need to supply credentials for the other feed. That way I can remove the ignore.

^ knowing this solves the main issue I have and means I can avoid the ignore arg.


2) if another project uses the feed. How do I specify the credentials for the feed in. .net core project? I’m googling but I can’t find what I’m looking for for this one. 

This I have solved for our current non core projects by specifying the .nuget folder with target and config. The legacy nuget restore on build for VS. 

But no idea how to solve this for new projects which aren’t in VS.

Thanks for your patients with my issue!

**Comment by nkolev92** [comment]
> When we built the project the second time it failed because it looked at our internal feed.

NuGet will try to use the global packages folder versions first. There's likely a version mismatch (badly authored package, non-exact dependencies etc). 
If that didn't satisfy the resolution logic, then it will go online and check all sources configure concurrently. 

> How can I specify the feeds for a single project so I don’t need to supply credentials for the other feed. That way I can remove the ignore.

There are a few ways to do that but each has it's problems.
1. Project level config, you add a config on the project level, you add the clear tag at the top of the sources and it will use only thing. Caveat here is this doesn't work in Visual Studio and there are no immediate plans to address it. https://github.com/NuGet/Home/issues/6746
1. RestoreSources property in the project file. 
Caveat here is that you can't provide credentials. There was also a bug there that is fixed in the 15.8 previews. https://github.com/NuGet/Home/issues/6776

> if another project uses the feed. How do I specify the credentials for the feed in. .net core project? I’m googling but I can’t find what I’m looking for for this one.

You can use the NuGet config to specify encrypted credentials in Visual Studio and nuget.exe or clear text in dotnet.exe(obv not recommended). 

The recommended approach would be to just acquire the credentials via Credential providers in VS and use the credential providers for nuget.exe and eventually (in 1.5 months or so) use the cross platform credential providers for dotnet.exe and msbuild https://github.com/NuGet/Home/issues/6486. 

> This I have solved for our current non core projects by specifying the .nuget folder with target and config. The legacy nuget restore on build for VS.

Can you clarify what you mean here? When you say non core projects, are those projects PackageReference or packages.config? 

Please note that in an ideal world, all projects in a repo would use the same feeds and you'd have only 1 config.  Similar to what we do [here](https://github.com/NuGet/NuGet.Client/blob/dev/NuGet.Config)

**Comment by phillip-haydon** [comment]
Yeah sorry typing on my phone. Project = Solution. I just meant project as application or all projects in the solution.

The Visual Studio thing was the old Enable Package Restore feature which added a .nuget folder which I can then add all the feeds to.

I’ll play around next week with amending our project based on the info you’ve given me. Prob write up a blog post too. Hopefully can resolve it and remove the ignore flag.

**Comment by nkolev92** [comment]
> The Visual Studio thing was the old Enable Package Restore feature which added a .nuget folder which I can then add all the feeds to.

You can simply have 1 config file with sources in the same folder as the solution. 

If possible, I'd suggest you simply have that 1 config and just authenticate for that feed. 
The http communication layer is shared among projects. Basically if you need a feed for 1 project, there is no harm in having it for a different one.

**Comment by nkolev92** [comment]
Hey @phillip-haydon 
I'm just closing the loop on this specific issue. 
Happy to help if you have any further questions.

**Comment by mrpubnight** [comment]
> Oh, I missed the log snippets, sorry about that.
> 
> As I mentioned above. All feeds have to be accessible in PackageReference for a restore to happen. This is by design as having different sources available can lead to drastically different resolution graphs.
> 
> The logs seem to indicate that you got a 401 from the private feed.
> 
> The problem is not that the package is not available on the feed. Rather it's, I can't connect to the feed. A feed that a user can't connect to is considered unreachable, hence fails the restore.
> 
> Next, NuGet firstly looks at the global packages folder (it's really an install location, I wouldn't call it a cache), before it goes to the online sources. If those packages are available in the global packages folder, it will use them. This is not as straightforward because NuGet will go online if there are imperfect dependency graphs or inexact versions. Namely "*" will cause NuGet to go online on the first restore. "2.0.0" while only having "2.5.0" in the version will do the same.
> 
> Please refer to the below doc for information on what happens when NuGet installs a package. https://docs.microsoft.com/en-us/nuget/consume-packages/ways-to-install-a-package#what-happens-when-a-package-is-installed
> 
> tl;dr;
> 
> I think the original assessment still stands.
> 
> All your sources need to available and this is by design.
> 
> Does this make sense?

Not trying to reopen this but wondering why NuGet appears to be looking at other sources if/when it is available in cache?  See  output logs below.

![image](https://user-images.githubusercontent.com/50109328/176707281-ce270bdb-49d1-4b6f-b955-71cef40ad898.png)
